### PR TITLE
8252219: C2: Randomize IGVN worklist for stress testing

### PR DESCRIPTION
**Pull request by robcasloz**
Add `StressIGVN` option to let C2 randomize IGVN worklist order. When enabled, the worklist is shuffled before each main run of the IGVN loop. Also add `GenerateStressSeed` and `StressSeed=N` options to randomly generate or specify the seed. In either case, the seed is logged if `LogCompilation` is enabled. The new options are declared as production+diagnostic for consistency with the existing `StressLCM` and `StressGCM` options.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252219](https://bugs.openjdk.java.net/browse/JDK-8252219): C2: Randomize IGVN worklist for stress testing

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to 4ae51be7cd4b10e1f2d772dd3616acf34c63b86b
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 4ae51be7cd4b10e1f2d772dd3616acf34c63b86b

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/242/head:pull/242`
`$ git checkout pull/242`

### CODE REVIEW

**Comment by robcasloz** [comment]
/summary
Add 'StressIGVN' option to let C2 randomize IGVN worklist order. When enabled,
the worklist is shuffled before each main run of the IGVN loop. Also add
'GenerateStressSeed' and 'StressSeed=N' options to randomly generate or specify
the seed. In either case, the seed is logged if 'LogCompilation' is enabled.
    
The generation or specification of seeds also affects the randomization
triggered by 'StressLCM' and 'StressGCM'. The new options are declared as
production+diagnostic for consistency with these existing options.

**Comment by robcasloz** [comment]
Reverted to "draft mode", as I just realized the design is not repeatable since it relies on global PRNG state.

**Comment by robcasloz** [comment]
/summary
Add 'StressIGVN' option to let C2 randomize IGVN worklist order. When enabled,
the worklist is shuffled before each main run of the IGVN loop. Also add
'GenerateStressSeed' and 'StressSeed=N' options to randomly generate or specify
the seed. In either case, the seed is logged if 'LogCompilation' is enabled.
The new options are declared as production+diagnostic for consistency with the
existing 'StressLCM' and 'StressGCM' options.

**Comment by robcasloz** [comment]
This pull request is ready for review again.

File: src/hotspot/share/opto/compile.cpp [hunk 1]
```diff
@@ -4436,8 +4450,21 @@ void Compile::remove_speculative_types(PhaseIterGVN &igvn) {
   }
 }
 
-// Auxiliary method to support randomized stressing/fuzzing.
-//
+// Auxiliary methods to support randomized stressing/fuzzing.
+
+int Compile::random() {
+  _stress_seed = os::next_random(_stress_seed);
```
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/hotspot/share/opto/compile.cpp:4456]: It is declared in this PR's version of src/hotspot/share/runtime/os.hpp, line 759:
https://github.com/openjdk/jdk/blob/e11318520bc548fae1a534dccc7e07218279f15b/src/hotspot/share/runtime/os.hpp#L759

**Comment by robcasloz** [comment]
Thanks for reviewing, Vladimir! I just addressed your two comments.

**Review by vnkozlov — APPROVED** [review]
Nice!

Did you try to run mach5 testing with IGV stress enabled by using --jvm-args "" mach5 option?

**Comment by robcasloz** [comment]
> Nice!
> 
> Did you try to run mach5 testing with IGV stress enabled by using --jvm-args "" mach5 option?

Thanks Vladimir! I tried something similar (but hackier) on tier1 and all tests passed, will try a more systematic run with multiple seeds, etc.

File: src/hotspot/share/opto/compile.cpp [hunk 2]
```diff
@@ -4436,8 +4450,21 @@ void Compile::remove_speculative_types(PhaseIterGVN &igvn) {
...
+  return static_cast<int>(_stress_seed);
+}
+
+void Compile::shuffle(Unique_Node_List* l) {
+  if (l->size() < 2) return;
+  for (uint i = l->size() - 1; i >= 1; i--) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/compile.cpp:4462]: You can remove the if-check as the loop check already covers it (loop is only executed if size >= 2).

**Review by chhagedorn — CHANGES_REQUESTED** [review]
Maybe you could add an additional HelloWorld test which only runs with your new flags to sanity check them without any other flags.

File: src/hotspot/share/opto/c2_globals.hpp [hunk 3]
```diff
@@ -49,6 +49,16 @@
   product(bool, StressGCM, false, DIAGNOSTIC,                               \
           "Randomize instruction scheduling in GCM")                        \
                                                                             \
+  product(bool, StressIGVN, false, DIAGNOSTIC,                              \
+          "Randomize worklist traversal in IGVN")                           \
+                                                                            \
+  product(bool, GenerateStressSeed, false, DIAGNOSTIC,                      \
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/c2_globals.hpp:55]: Is this flag really required? We could simply generate the seed if StressSeed has not been specified on the command line (see `FLAG_IS_DEFAULT` macro).

File: src/hotspot/share/opto/node.cpp [hunk 4]
```diff
@@ -2329,6 +2329,13 @@ void Node_Array::sort( C_sort_func_t func) {
   qsort( _nodes, _max, sizeof( Node* ), func );
 }
 
+//-----------------------------------------------------------------------------
+void Node_Array::swap(uint i, uint j) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/node.cpp:2333]: Use can use the swap method from globalDefinitions.hpp.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
We should think about adding the Stress* flags to some tier in the CI.

File: src/hotspot/share/opto/phaseX.cpp [hunk 5]
```diff
@@ -1150,6 +1150,7 @@ void PhaseIterGVN::trace_PhaseIterGVN_verbose(Node* n, int num_processed) {
 void PhaseIterGVN::optimize() {
   DEBUG_ONLY(uint num_processed  = 0;)
   NOT_PRODUCT(init_verifyPhaseIterGVN();)
+  if (StressIGVN) C->shuffle(&_worklist);
```
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/hotspot/share/opto/phaseX.cpp]: Done.

**Comment by robcasloz** [comment]
> Maybe you could add an additional HelloWorld test which only runs with your new flags to sanity check them without any other flags.

Good idea, I just did that.

File: src/hotspot/share/opto/compile.cpp (hunk 2 above)
Comment (robcasloz) [reply src/hotspot/share/opto/compile.cpp]: Note that `size()` could be 0, leading `l->size() - 1` to underflow.

File: src/hotspot/share/opto/c2_globals.hpp (hunk 3 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/c2_globals.hpp:55]: Agree.

File: src/hotspot/share/opto/compile.cpp (hunk 2 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/compile.cpp]: You're right, it's `uint` and not `int` - then you can leave it as it is.

File: src/hotspot/share/opto/c2_globals.hpp (hunk 3 above)
Comment (robcasloz) [reply src/hotspot/share/opto/c2_globals.hpp]: Done. I didn't know about `FLAG_IS_DEFAULT`, thanks!

File: src/hotspot/share/opto/node.cpp (hunk 4 above)
Comment (robcasloz) [reply src/hotspot/share/opto/node.cpp]: Done, thanks! I also moved swapping to `PhaseIterGVN` to simplify the PR.

**Comment by robcasloz** [comment]
/summary
Add 'StressIGVN' option to let C2 randomize IGVN worklist order. When enabled,
the worklist is shuffled before each main run of the IGVN loop. Also add
'StressSeed=N' option to specify the seed. If the seed is not specified, a
random one is generated. In either case, the seed is logged if 'LogCompilation'
is enabled. The new options are declared as production+diagnostic for
consistency with the existing 'StressLCM' and 'StressGCM' options.

**Review by TobiHartmann — APPROVED** [review]
Thanks for making these changes. Looks good to me.

**Review by chhagedorn — APPROVED** [review]
Looks good to me! Thanks for adding the additional test.

File: test/hotspot/jtreg/compiler/arguments/TestStressIGVNOptions.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+/*
+ * @test
+ * @bug 8252219
+ * @summary Tests that different combinations of the options -XX:+StressIGVN and
+ *          -XX:StressSeed=N are accepted.
+ * @run main/othervm -XX:+StressIGVN
+ *      compiler.arguments.TestStressIGVNOptions
+ * @run main/othervm -XX:+StressIGVN -XX:StressSeed=42
+ *      compiler.arguments.TestStressIGVNOptions
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/arguments/TestStressIGVNOptions.java:32]: Please, add next to run test when C2 is enabled as you did for other 2 tests:
@requires vm.compiler2.enabled

Comment (robcasloz) [reply test/hotspot/jtreg/compiler/arguments/TestStressIGVNOptions.java:33]: Done, thanks for catching this up!

**Comment by robcasloz** [comment]
> > Nice!
> > Did you try to run mach5 testing with IGV stress enabled by using --jvm-args "" mach5 option?
> 
> Thanks Vladimir! I tried something similar (but hackier) on tier1 and all tests passed, will try a more systematic run with multiple seeds, etc.

I ran `tier1_compiler` with `-XX:+StressIGVN` for 100 repetitions and did not trigger any failure.

**Review by vnkozlov — APPROVED** [review]

**Pull request closed without merging**